Bridge documentation (#5705)
Particular/docs.particular.net: nservicebus/bridge/migrating-from-router.md

Because of this, the NServiceBus.Router must remain running until all in-flight messages have been delivered. The time this takes is highly individual to each system. If it is unclear what the right time is to stop the NServiceBus.Router, contact [Particular Software](https://particular.net/contactus) for guidance.

#### Commands

Commands are the easiest to migrate as they have a specific route to a specific logical endpoint. After the transport bridge has been set up and running, every endpoint can remove any NServiceBus.Router-related configuration for routing commands using the connector. This implies that the original routing towards the specific endpoint can be restored and the bridge will make sure it is transferred to the correct transport.

#### Events

> Events are the hardest message type to migrate to the transport bridge from NServiceBus.Router.
> 
> The issue is with the subscriptions that are registered within a message broker. The scenario could exist as seen in the graph below, where `Endpoint B` publishes an event and the event is sent to both the `NServiceBus.Router` and the `NServiceBus.Transport.Bridge`. As a result, `Endpoint C` could receive the same event twice.

Reviewer: ```suggestion
Events require more effort to migrate to the transport bridge from NServiceBus.Router. The issue is with the subscriptions that are registered within a message broker. The scenario could exist as seen in the graph below, where `Endpoint B` publishes an event and the event is sent to both the `NServiceBus.Router` and the `NServiceBus.Transport.Bridge`. As a result, `Endpoint C` could receive the same event twice.
```

Note that both events would have the same message identifier, which allows the [outbox](/nservicebus/outbox/) to de-duplicate the messages. Without the outbox, either the message handler must be [idempotent](/nservicebus/concepts/glossary.md#idempotence) or the event will be processed twice.

```mermaid
flowchart LR
  Br(Bridge)
  A[Endpoint A] <---> Br
  B[Endpoint B] <---> Br
  Br <---> C[Endpoint C]
  Br <---> D[Endpoint D]
  Ro[Router] <---> C
  B <--> Ro
  subgraph msmq
    A
    B
  end
  subgraph sql
    C
    D
  end
```

For this scenario, take the following steps to move from NServiceBus.Router to NServiceBus.Transport.Bridge:

> 1. Take down both `EndpointB` and the `Bridge`.

Reviewer: ```suggestion
1. Stop both `EndpointB` and the `Bridge`.
```

1. Configure the `Bridge` to register a publisher with `Endpoint C` for the event that `Endpoint B` publishes.
2. Remove the subscription from the message broker for the `Router`.
3. Add the subscription to the message broker for the `Bridge`.
4. Start the `Bridge`.
5. Start `Endpoint B`.

> When the `Bridge` starts up, it will also create the subscription; it is not required to manually create the subscription in the message broker. However, tt is important to be aware of when `Endpoint B` cannot be taken offline due to high availability, scaling out or service level agreements.

Reviewer: ```suggestion
When the `Bridge` starts up, it will also create the subscription; it is not required to manually create the subscription in the message broker. However, be aware of when `Endpoint B` cannot be taken offline due to high availability, scaling out or service level agreements.
```

NOTE: If Endpoint B cannot be taken offline, the time between step 3 and step 4 determines how many events will be lost.

#### Additional messages

> There may be additional messages, like request/response messages, which route a message back to the endpoint it originated from. These must be routed back using the NServiceBus.Router component. The result is that as long as there are messages like these to be processed, the NServiceBus.Router component must be active within the system.

Reviewer: ```suggestion
There may be additional messages, like request/response messages, which route a message back to the endpoint it originated from. These must be routed back using the NServiceBus.Router component. The result is that as long as there are messages like these to be processed, the NServiceBus.Router component must be active within the system alongside the bridge.
```

## Conclusion

> Migrating from NServiceBus.Router to NServiceBus.Transport.Bridge is not difficult, but it is something that requires consideration and analysis before the NServiceBus.Router component can be completely removed from a system.

Reviewer: ```suggestion
Migrating from NServiceBus.Router to NServiceBus.Transport.Bridge is not difficult, but it requires careful consideration and analysis before the NServiceBus.Router component can be completely removed from a system.
```
